### PR TITLE
test out validation plugin

### PR DESCRIPTION
**Pull request by FriederikeHanssen**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: assets/schema_input.json [hunk 1]
```diff
@@ -31,30 +35,19 @@
...
-                        "maxLength": 0
-                    }
-                ]
+                "type": "string",
+                "pattern": "^\\S+$",
+                "unique": ["patient", "sample"],
+                "dependentRequired": ["fastq_1"],
+                "meta": ["lane"]
             },
             "fastq_1": {
```
Comment (maxulysse) [inline assets/schema_input.json:52]: can we add patterns to ensure it finishes with `{fastq,fq}.gz`?

File: assets/schema_input.json [hunk 2]
```diff
@@ -80,7 +75,9 @@
                         "type": "string",
                         "maxLength": 0
                     }
-                ]
+                ],
+                "format": "file-path",
+                "exists": true
             },
             "table": {
```
Comment (maxulysse) [inline assets/schema_input.json:82]: can we add patterns to ensure it finishes with `.table`?

File: assets/schema_input.json (hunk 1 above)
Comment (maxulysse) [reply assets/schema_input.json:52]: forget I said anything

File: main.nf [hunk 3]
```diff
@@ -68,6 +68,22 @@ params.vep_version           = WorkflowMain.getGenomeAttribute(params, 'vep_vers
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */
 
+include { validateParameters; paramsHelp } from 'plugin/nf-validation'
+
+// Print help message if needed
+if (params.help) {
+    def logo = NfcoreTemplate.logo(workflow, params.monochrome_logs)
+    def citation = '\n' + WorkflowMain.citation(workflow) + '\n'
+    def String command = "nextflow run ${workflow.manifest.name} --input samplesheet.csv --genome GRCh38 -profile docker"
```
Comment (maxulysse) [inline main.nf:77]:
```suggestion
    def String command = "nextflow run ${workflow.manifest.name} --input samplesheet.csv --genome GATK.GRCh38 -profile docker"
```

Comment (maxulysse) [reply main.nf:77]: shouldn't the default command contain --outdir as well?

File: nextflow.config [hunk 4]
```diff
@@ -125,6 +122,13 @@ params {
     max_memory                 = '128.GB'
     max_cpus                   = 16
     max_time                   = '240.h'
+
+    // Schema validation default options
+    validationFailUnrecognisedParams = false
+    validationLenientMode            = false
+    validationSchemaIgnoreParams     = 'genomes,snpeff_version,vep_version,cf_ploidy'
+    validationShowHiddenParams       = false
+    validate_params                  = true
```
Comment (maxulysse) [inline nextflow.config:131]: it's not `validateParams`?
I don't follow this new pattern for the params :rofl:

File: assets/schema_input.json (hunk 1 above)
Comment (mirpedrol) [reply assets/schema_input.json:52]: we could add it if you need it

File: nextflow.config (hunk 4 above)
Comment (mirpedrol) [reply nextflow.config:131]: `validate_params` is an nf-core param, not an nf-validation/nextflow one

File: workflows/sarek.nf [hunk 5]
```diff
@@ -56,7 +63,129 @@ def checkPathParamList = [
...
+        def CN         = params.seq_center ? "CN:${params.seq_center}\\t" : ''
+
+        def flowcell   = flowcellLaneFromFastq(fastq_1)
+        // Don't use a random element for ID, it breaks resuming
+        def read_group = "\"@RG\\tID:${flowcell}.${meta.sample}.${meta.lane}\\t${CN}PU:${meta.lane}\\tSM:${meta.patient}_${meta.sample}\\tLB:${meta.sample}\\tDS:${params.fasta}\\tPL:${params.seq_platform}\""
+        println "Number of lanes: $num_lanes"
+
+        meta           = meta + [num_lanes: num_lanes.toInteger(), read_group: read_group.toString(), data_type: 'fastq', size: 1]
+
+        if (params.step == 'mapping') return [ meta - meta.subMap('lane'), [ fastq_1, fastq_2 ] ]
```
Comment (maxulysse) [inline workflows/sarek.nf:89]: @FriederikeHanssen don't we need meta.lane at this step still?

File: workflows/sarek.nf [hunk 6]
```diff
@@ -56,7 +63,129 @@ def checkPathParamList = [
...
+ch_from_samplesheet = params.build_only_index ? Channel.empty() : Channel.fromSamplesheet("input")
+
+ch_from_samplesheet
+.map{ meta, fastq_1, fastq_2, table, cram, crai, bam, bai, vcf, variantcaller ->
+    [ meta.patient + meta.sample, [meta, fastq_1, fastq_2, table, cram, crai, bam, bai, vcf, variantcaller] ]
+}.tap{ ch_with_patient_sample }
+.groupTuple()
+.map { patient_sample, ch_items ->
+    [ patient_sample, ch_items.size() ]
+}.combine(ch_with_patient_sample, by: 0)
```
Comment (maxulysse) [inline workflows/sarek.nf:75]: :exploding_head:

File: main.nf (hunk 3 above)
Comment (mirpedrol) [reply main.nf:77]: It's not there [now](https://github.com/nf-core/sarek/blob/master/lib/WorkflowMain.groovy#L27C71-L27C82), but it's true we should add it as it is required 👍

File: workflows/sarek.nf [hunk 7]
```diff
@@ -56,7 +63,129 @@ def checkPathParamList = [
 for (param in checkPathParamList) if (param) file(param, checkIfExists: true)
 
 // Set input, can either be from --input or from automatic retrieval in WorkflowSarek.groovy
-input_sample = params.build_only_index ? Channel.empty() : extract_csv(file(WorkflowSarek.retrieveInput(params, log), checkIfExists: true))
+ch_from_samplesheet = params.build_only_index ? Channel.empty() : Channel.fromSamplesheet("input")
```
Comment (maxulysse) [inline workflows/sarek.nf:66]: how do you deal with the automatic restart from another step with the csv that are produced by sarek when no input specified?

Comment (mirpedrol) [reply workflows/sarek.nf:66]: 🤔 I don't think I modified this code, but do you have a test dataset for this, so we can be sure?

**Comment by maxulysse** [comment]
@mirpedrol we just merged in the template sync

File: workflows/sarek.nf (hunk 7 above)
Comment (maxulysse) [reply workflows/sarek.nf:66]: regular `-profile test` or `-profile test_cache`
run it once, and then run it again with --input=false and any --step other than mapping.
so probably --step variant_calling for a quick test

File: nextflow_schema.json [hunk 8]
```diff
@@ -14,14 +14,22 @@
...
+                    "anyOf": [
+                        {
+                            "type": "string",
+                            "format": "file-path",
+                            "exists": true,
+                            "mimetype": "text/csv",
+                            "pattern": "^\\S+\\.csv$"
+                        },
+                        {
+                            "type": "boolean"
```
Comment (maxulysse) [inline nextflow_schema.json:30]: I'm guessing we can either have false or null here

Comment (maxulysse) [reply nextflow_schema.json:30]: true should not work, even if boolean

Comment (mirpedrol) [reply nextflow_schema.json]: I've changed this to allowing an empty string instead of a boolean

File: main.nf [hunk 9]
```diff
@@ -62,6 +62,14 @@ params.vep_genome            = WorkflowMain.getGenomeAttribute(params, 'vep_geno
 params.vep_species           = WorkflowMain.getGenomeAttribute(params, 'vep_species')
 params.vep_version           = WorkflowMain.getGenomeAttribute(params, 'vep_version')
 
+/*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    ALTERNATIVE INPUT FILE ON RESTART
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+params.input_restart = WorkflowSarek.retrieveInput(params, log)
```
Comment (mirpedrol) [inline main.nf:71]: Here we define the new parameter `input_restart` which will be used if input is an empty string

File: assets/schema_input.json [hunk 10]
```diff
@@ -10,15 +10,19 @@
             "patient": {
                 "type": "string",
                 "pattern": "^\\S+$",
-                "errorMessage": "ID name must be provided and cannot contain spaces"
+                "errorMessage": "Patient name must be provided and cannot contain spaces",
```
Comment (maxulysse) [inline assets/schema_input.json:13]:
```suggestion
                "errorMessage": "Patient ID must be provided and cannot contain spaces",
```

File: assets/schema_input.json [hunk 11]
```diff
@@ -10,15 +10,19 @@
...
+                "meta": ["patient"]
             },
             "sample": {
                 "type": "string",
                 "pattern": "^\\S+$",
-                "errorMessage": "Sample name must be provided and cannot contain spaces"
+                "errorMessage": "Sample name must be provided and cannot contain spaces",
```
Comment (maxulysse) [inline assets/schema_input.json:19]:
```suggestion
                "errorMessage": "Sample ID must be provided and cannot contain spaces",
```

File: assets/schema_input.json (hunk 10 above)
Comment (maxulysse) [reply assets/schema_input.json:13]: Just me being paranoid, I know we all know that we don't want the actual name of patient here, but I'd rather we cover all our bases.
I'm even concerned that patient ID might be misinterpreted...

Comment (mirpedrol) [reply assets/schema_input.json:13]: Makes sense! Can also go back to "ID name", I didn't know the reason when I changed it

File: lib/WorkflowSarek.groovy [hunk 12]
```diff
@@ -119,9 +119,8 @@ class WorkflowSarek {
...
 
     public static String retrieveInput(params, log){
-        def input = ''
-        if (params.input) input = params.input
-        else {
+        def input = null
+        if (!params.input && !params.build_only_index) {
             switch (params.step) {
                 case 'mapping':                 Nextflow.error("Can't start with step $params.step without samplesheet")
                                                 break
```
[the remark this answers is not in the document]
Comment (mirpedrol) [reply lib/WorkflowSarek.groovy:126]: We use the variable `input`, I didn't change the name, but we set `params.input_restart` when we call the function [here](https://github.com/nf-core/sarek/pull/1059/files/c18ea1280ab0b660a93980e3ffc403cc72c4dcf0#diff-6401496ba455b9488ffa902a6e4d7732b2c60ff2d77c5c3ef96b28a7ac7d3b28R71).

File: workflows/sarek.nf [hunk 13]
```diff
@@ -66,10 +69,133 @@ def checkPathParamList = [
...
+.map { patient_sample, num_lanes, ch_items ->
+    (meta, fastq_1, fastq_2, table, cram, crai, bam, bai, vcf, variantcaller) = ch_items
+    if (meta.lane && fastq_2) {
+        meta           = meta + [id: "${meta.sample}-${meta.lane}".toString()]
+        def CN         = params.seq_center ? "CN:${params.seq_center}\\t" : ''
+
+        def flowcell   = flowcellLaneFromFastq(fastq_1)
+        // Don't use a random element for ID, it breaks resuming
+        def read_group = "\"@RG\\tID:${flowcell}.${meta.sample}.${meta.lane}\\t${CN}PU:${meta.lane}\\tSM:${meta.patient}_${meta.sample}\\tLB:${meta.sample}\\tDS:${params.fasta}\\tPL:${params.seq_platform}\""
+        println "Number of lanes: $num_lanes"
```
[the remark this answers is not in the document]
Comment (mirpedrol) [reply workflows/sarek.nf:97]: Shouldn't be printed :D
```suggestion
```

File: workflows/sarek.nf [hunk 14]
```diff
@@ -66,10 +69,133 @@ def checkPathParamList = [
...
+
+ch_from_samplesheet
+.map{ meta, fastq_1, fastq_2, table, cram, crai, bam, bai, vcf, variantcaller ->
+    [ meta.patient + meta.sample, [meta, fastq_1, fastq_2, table, cram, crai, bam, bai, vcf, variantcaller] ]
+}.tap{ ch_with_patient_sample }
+.groupTuple()
+.map { patient_sample, ch_items ->
+    [ patient_sample, ch_items.size() ]
+}.combine(ch_with_patient_sample, by: 0)
+.map { patient_sample, num_lanes, ch_items ->
```
[the remark this answers is not in the document]
Comment (mirpedrol) [reply workflows/sarek.nf:88]: We group the channel by patient+sample with `groupTuple()`, then we obtain the size, the number of channels which have the same `groupTubple()` with `ch_items.size()`

File: workflows/sarek.nf (hunk 5 above)
Comment (mirpedrol) [reply workflows/sarek.nf:101]:
```suggestion
        if (params.step == 'mapping') return [ meta, [ fastq_1, fastq_2 ] ]
```

**Review by maxulysse — APPROVED** [review]

File: workflows/sarek.nf [hunk 15]
```diff
@@ -66,10 +69,132 @@ def checkPathParamList = [
...
+
+    // start from BAM
+    } else if (meta.lane && bam) {
+        if (params.step != 'mapping' && !bai) {
+            error("BAM index (bai) should be provided.")
+        }
+        meta            = meta + [id: "${meta.sample}-${meta.lane}".toString()]
+        def CN          = params.seq_center ? "CN:${params.seq_center}\\t" : ''
+        def read_group  = "\"@RG\\tID:${meta.sample}_${meta.lane}\\t${CN}PU:${meta.lane}\\tSM:${meta.patient}_${meta.sample}\\tLB:${meta.sample}\\tDS:${params.fasta}\\tPL:${params.seq_platform}\""
+        println "Number of lanes: $num_lanes"
```
Comment (mirpedrol) [inline workflows/sarek.nf:113]:
```suggestion
```

**Pull request merged**